### ISSUE TITLE
SC2269: Add self-assignment of array elements to checks

### ISSUE THREAD

**Issue by Simon-Brandt** [koalaman/shellcheck#3150]
#### For bugs

- Rule Id (if any, e.g. SC1000): SC2269
- My shellcheck version (`shellcheck --version` or 'online'): online
- [X] I tried on shellcheck.net and verified that this is still a problem on the latest commit
- [ ] It's not reproducible on shellcheck.net, but I think that's because it's an OS, configuration or encoding issue

#### For new checks and feature suggestions

- [X] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [X] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```bash
#!/bin/bash
# SC2269 not triggered for indexed arrays
array=(1 2 3 4)
for i in "${!array[@]}"; do
    array[i]="${array[i]}"
done
```

or likewise:

```bash
#!/bin/bash
# SC2269 not triggered for associative arrays
declare -A array
array[foo]=1
array[bar]=2
array[baz]=3

for k in "${!array[@]}"; do
    array[k]="${array[k]}"
done
```

Contrast this with

```bash
#!/bin/bash
# SC2269 triggered for simple variable assignments
foo=1
foo="${foo}"
```

#### Here's what shellcheck (https://www.shellcheck.net/) currently says:

```
No issues detected!
```

(both for indexed and associative arrays)

#### Here's what I wanted or expected to see:

```
^-- [SC2269](https://www.shellcheck.net/wiki/SC2269) (info): This variable is assigned to itself, so the assignment does nothing.
```

(as output for the third example with normal variable assignments)

Reason: Assigning array elements to themselves doesn't change their value, just as assigning variables to themselves doesn't change them. Still, for arrays, this may be more involved to check due to the iteration (and not even counting the possible indirect references that ShellCheck cannot see, as in [SC2302](https://www.shellcheck.net/wiki/SC2302)).

As far as I see, the respective check is done in:

https://github.com/koalaman/shellcheck/blob/d3001f337aa3f7653a621b302261f4eac01890d0/src/ShellCheck/Analytics.hs#L4259-L4277

Might it be possible to extend the checks to array elements, for shells supporting them?
